### PR TITLE
Support for catalog refreshing dynamically by changing catalog configuration files

### PR DESCRIPTION
**Pull request by liyubin117**
Based on offical StaticCatalogStore, I have Inherited the class to support catalog refreshing dynamically, launch a seperated thread to sense file changes of catalog directory via java.nio.WatchService, and then update data source announcement.
otherwise, support customize wheter to enable the feature using etc/node.properties, default is disabled:
```
catalog.dynamic-update.enabled = true
```

```
== RELEASE NOTES ==

General Changes
* Support for catalog refreshing dynamically by changing catalog configuration files

```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/metadata/DynamicCatalogStore.java [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    }
+
+    public void loadCatalogs()
+            throws Exception
+    {
+        super.loadCatalogs();
+
+        if (dynamicUpdateEanbled) {
+            // update catalogs dynamically
+            new Thread(() -> {
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/metadata/DynamicCatalogStore.java]: Use executors than new a Thread

File: presto-main/src/main/java/com/facebook/presto/metadata/DynamicCatalogStore.java [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+import java.nio.file.Paths;
+import java.nio.file.StandardWatchEventKinds;
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.util.concurrent.TimeUnit;
+
+import static com.facebook.presto.server.PrestoServer.updateDatasourcesWithSpecifiedAction;
+
+public class DynamicCatalogStore
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/metadata/DynamicCatalogStore.java]: The inheritance conflicts the naming. Separate them apart or rename `StaticCatalogStore`

File: presto-main/src/main/java/com/facebook/presto/metadata/DynamicCatalogStore.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+            }
+        }
+    }
+
+    private void addCatalog(Path catalogPath)
+    {
+        File file = new File(getCatalogConfigurationDir(), catalogPath.getFileName().toString());
+        if (file.isFile() && file.getName().endsWith(".properties")) {
+            try {
+                TimeUnit.SECONDS.sleep((long) watchTimeout.getValue(TimeUnit.SECONDS));
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/metadata/DynamicCatalogStore.java]: Use future/executor/threadpool. Don't block the thread

File: presto-main/src/main/java/com/facebook/presto/server/PrestoServer.java [hunk 4]
```diff
@@ -77,6 +77,12 @@ public static void main(String[] args)
         new PrestoServer().run();
     }
 
+    private static Announcer announcer;
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/server/PrestoServer.java]: this can be a local var

**Comment by liyubin117** [comment]
> Skimmed through the patch. What is the use case behind this? Is it really necessary to add/drop catalogs at runtime? Pass to @wenleix to comment

The feature could make it possible to configure catalog and then see the effect in time,  Improve the development efficiency of the test environment at least, In face, the feature is a strong demand in our application scenario, we have applied it in our production environment for two months :).

**Pull request closed without merging**